arch/x86/kvm/pkvm/pkvm.c

```diff
@@ -475,14 +475,24 @@ static struct pkvm_vcpu *load_pkvm_vcpu(struct kvm_vcpu *shared_vcpu, unsigned l
 		return NULL;
 
 	vcpu = to_kvm_vcpu(pkvm_vcpu);
-	if (!is_kvm_vcpu_accessible(vcpu, fn)) {
-		put_pkvm_vcpu(pkvm_vcpu);
-		return NULL;
+	if (!is_kvm_vcpu_accessible(vcpu, fn))
+		goto out;
+
+	/*
+	 * Except the vcpu_load PV interface, the other vcpu accessing PV
+	 * interfaces requires the vcpu to be loaded on this CPU.
+	 */
+	if (vcpu->cpu != raw_smp_processor_id()) {
+		if (vcpu->cpu != -1 || fn != __pkvm__vcpu_load)
```

Comment:
If we allow `vcpu->cpu == -1` (i.e. allow the host to do `vcpu_run` and other PV ops without doing `vcpu_load` first), it means the host can do some PV ops for the same vCPU on two different pCPUs simultaneously (as long as it hasn't vcpu_load'ed this vCPU on any pCPU), causing races in pKVM?

Why exactly do we need to allow `vcpu->cpu == -1`?

Reply:
Also nit: this can be in one `if` instead of two?

Reply:
> If we allow `vcpu->cpu == -1` (i.e. allow the host to do `vcpu_run` and other PV ops without doing `vcpu_load` first), it means the host can do some PV ops for the same vCPU on two different pCPUs simultaneously (as long as it hasn't vcpu_load'ed this vCPU on any pCPU), causing races in pKVM?

Let me try to understand this scenario better, the scenario is the host use vcpu_run PV interface to run a vcpu which is not loaded first, is this correct? If so, then vcpu->cpu == -1, so the 2nd if will check fn. As it is __pkvm_vcpu_run PV interface, not __pkvm_vcpu_load, so the 2nd if is true, and goes out, which returns NULL. This seems fine. No?

> 
> Why exactly do we need to allow `vcpu->cpu == -1`?

To allow the __pkvm_vcpu_load PV interface. This is the only PV interface so far has vcpu->cpu == -1.
So we only allow the host to do __pkvm_vcpu_load for the same vcpu on two different pCPUs simultaneously. And such race will be prevented by the other patch via the cmpxchg.

Reply:
> Also nit: this can be in one `if` instead of two?

Yes, it can be in one `if`, just a bit longer line.
Probably can write it like this, which seems easier to for reading?

```suggestion
        if (vcpu->cpu == raw_smp_processor_id() ||
            (vcpu->cpu == -1 && fn == __pkvm__vcpu_load)) {
                pkvm_x86_call(switch_to_guest_vcpu)(vcpu);
                return pkvm_vcpu;
        }
```

Reply:
Yeah, I got confused, I thought this condition allows the host to run a vcpu  which is not loaded, but now I see it doesn't allow that. It would be easier for me to get that if it was written as `!(vcpu->cpu == -1 && fn == __pkvm__vcpu_load)` (to clearly emphasize that __pkvm__vcpu_load is the special case when cpu == -1 is allowed, while in all other cases it is not). :)

Reply:
BTW also just a thought: maybe it would be cleaner to not call `pkvm_x86_call(switch_to_guest_vcpu)(vcpu)` if fn is __pkvm__vcpu_load?

If so, then in in vmx_switch_to_guest_vcpu() we could remove the check `vmx->loaded_vmcs->cpu == -1`, and then the check `WARN_ON_ONCE(vmx->loaded_vmcs->cpu != cpu)` would cover more cases?

Reply:
> Yeah, I got confused, I thought this condition allows the host to run a vcpu which is not loaded, but now I see it doesn't allow that. It would be easier for me to get that if it was written as `!(vcpu->cpu == -1 && fn == __pkvm__vcpu_load)` (to clearly emphasize that __pkvm__vcpu_load is the special case when cpu == -1 is allowed, while in all other cases it is not). :)

Sure, I can update to make it easier to read.

Reply:
> BTW also just a thought: maybe it would be cleaner to not call `pkvm_x86_call(switch_to_guest_vcpu)(vcpu)` if fn is __pkvm__vcpu_load?
> 
> If so, then in in vmx_switch_to_guest_vcpu() we could remove the check `vmx->loaded_vmcs->cpu == -1`, and then the check `WARN_ON_ONCE(vmx->loaded_vmcs->cpu != cpu)` would cover more cases?

I initially want to make the `pkvm_x86_call(switch_to_guest_vcpu)(vcpu)` and `pkvm_x86_call(switch_to_host_vcpu)(vcpu)` in pair for easier reading. As the `pkvm_x86_call(switch_to_host_vcpu)(vcpu)` is called unconditionally, also made the `pkvm_x86_call(switch_to_guest_vcpu)(vcpu)` unconditionally. How do you think?

Reply:
> As the `pkvm_x86_call(switch_to_host_vcpu)(vcpu)` is called unconditionally, also made the `pkvm_x86_call(switch_to_guest_vcpu)(vcpu)` unconditionally.

Good point. So there are pros and cons. Ok, I guess we can keep it as is.

Reply:
> As the `pkvm_x86_call(switch_to_host_vcpu)(vcpu)` is called unconditionally, also made the `pkvm_x86_call(switch_to_guest_vcpu)(vcpu)` unconditionally.

Good point. So there are pros and cons. Ok, I guess we can keep it as is.